File: DSA/Hima-Swarupa-Padala/day20q1.py
def can_be_colored(vertices, edges):
    
    graph = {i: [] for i in range(1, vertices + 1)}
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)

    colors = {}

    def dfs(vertex, color):
        colors[vertex] = color
        for neighbor in graph[vertex]:
            if neighbor not in colors:
                if not dfs(neighbor, 3 - color):
                    return False
            elif colors[neighbor] == color:
                return False
        return True

    if not dfs(1, 1):
        return "NO"
    return "YES"

vertices = 4
edges = [(1, 2), (1, 3), (2, 4), (3, 4)]
print(can_be_colored(vertices, edges)) 

File: DSA/Yashvardhan-Guliyani/Day20_q1.java
import java.util.*;

public class GraphColoring {
    private int V;
    private List<List<Integer>> adj; 

    public GraphColoring(int v) {
        V = v;
        adj = new ArrayList<>(V);
        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }
    }

    public void addEdge(int v, int w) {
        adj.get(v).add(w);
        adj.get(w).add(v);
    }

    public boolean isBipartite() {
        int[] color = new int[V];
        Arrays.fill(color, -1);

        for (int i = 0; i < V; i++) {
            if (color[i] == -1) {
                if (!dfsColor(i, 0, color)) {
                    return false;
                }
            }
        }
        return true;
    }

    private boolean dfsColor(int v, int c, int[] color) {
        color[v] = c;

        for (int neighbor : adj.get(v)) {
            if (color[neighbor] == -1) {
                if (!dfsColor(neighbor, 1 - c, color)) {
                    return false;
                }
            } else if (color[neighbor] == color[v]) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Enter the number of vertices: ");
        int V = scanner.nextInt();
        
        System.out.print("Enter the number of edges: ");
        int E = scanner.nextInt();
        
        GraphColoring graph = new GraphColoring(V);
        
        System.out.println("Enter the edges (vertex pairs):");
        for (int i = 0; i < E; i++) {
            int v = scanner.nextInt() - 1; 
            int w = scanner.nextInt() - 1;
            graph.addEdge(v, w);
        }
        
        System.out.println(graph.isBipartite() ? "YES" : "NO");
        
        scanner.close();
    }
}
